include/mysqlpp/udf_wrappers.hpp

```diff
@@ -41,55 +43,27 @@ namespace mysqlpp {
 template <typename ImplType>
 struct udf_impl_meta_info;
 
-template <typename ImplType, item_result_type ItemResult>
-class generic_udf_base {
- public:
-  static bool init(UDF_INIT *initid, UDF_ARGS *args, char *message) noexcept {
-    udf_context udf_ctx{initid, args};
-    extended_impl_t *impl = nullptr;
-    try {
-      impl = new extended_impl_t{udf_ctx};
-    } catch (const std::exception &e) {
-      std::strncpy(message, e.what(), MYSQL_ERRMSG_SIZE);
-      message[MYSQL_ERRMSG_SIZE - 1] = '\0';
-      return true;
-    } catch (...) {
-      std::strncpy(message, "unexpected exception", MYSQL_ERRMSG_SIZE);
-      message[MYSQL_ERRMSG_SIZE - 1] = '\0';
-      return true;
-    }
-
-    initid->ptr = reinterpret_cast<char *>(impl);
-
-    return false;
-  }
-
-  static void deinit(UDF_INIT *initid) noexcept {
-    delete get_extended_impl_from_udf_initid(initid);
-  }
-
- protected:
-  using extended_impl_t = impl_with_mixin<udf_mixin_t<ItemResult>, ImplType>;
-
-  static extended_impl_t *get_extended_impl_from_udf_initid(
-      UDF_INIT *initid) noexcept {
-    return reinterpret_cast<extended_impl_t *>(initid->ptr);
-  }
-  static const char *get_function_label(std::string &buffer) noexcept {
+class udf_base {
```

Comment:
So this is non-parameterized part of generic_udf_base, isn't it? Could you please explain why it is better to keep it separate? Perhaps worth a comment?

Reply:
Yes, @dlenev
`udf_base` is basically a part of the `generic_udf_base` / `generic_udf` logic that does not depend on the template parameters. The main reason for doing this is to eliminate so called `c++ template code bloat` problem.
When we have something like
```c++
template<typename T> class primitive_vector {
public:
  ...
  T& at(std::size_t index) {
    validate_index();
    return data_[index];
  }
private:
  std::size_t size_;
  T *data_;

  void validate_index(std::size_t index) const {
    if (index >= size_) throw std::out_of_range{};
  }
};
```
For `primitive_vector<int>` and `primitive_vector<double>` we will have two instantiations of the `validate_index()` function with absolutely identical body:
```c++
primitive_vector<int>::validate_index
primitive_vector<double>::validate_index
```
This can drastically affect the size of the binary, linking time and even runtime performance (because of more instruction cache misses).
So, a typical solution is to extract a part of the logic that does not depend on the template parameter into a base class.
```c++
class primitive_vector_base {
public:
  ...
protected:
  std::size_t size_;
  void validate_index(std::size_t index) const {
    if (index >= size_) throw std::out_of_range{};
  }
};
template<typename T> class primitive_vector: private primitive_vector_base {
public:
  ...
  T& at(std::size_t index) {
    validate_index();
    return data_[index];
  }
private:
  T *data_;
};

```
BTW, in both STL implementations that I know (`libc++ ` and `libstdc++`) all the containers are built this way.

Reply:
Added a comment

Reply:
OK. Thank you! I know that MySQL List<> templates use the same approach to solve bloat problem but somehow missed that the same can apply here.